File: web/src/api/interpretation.ts
import { downloadArtifact, fetchArtifactBlob, request } from './client';
import type { ResourcePolicy } from './development';
import type { LifecycleState } from './lifecycle';
import type { ComputeExecution } from './predictors';
import type { Finding } from './scientific';
export interface InterpretationSlideInput {
  slideId: string; slidePath: string; featurePath: string; coordinatesPath?: string | null;
  featureKey: string; coordinatesKey: string; patchWidthLevel0?: number; patchHeightLevel0?: number;
  coordinateSpace: 'level0'; confirmRowAlignment: boolean;
}
export interface SlideGeometry { width: number; height: number; backend: string; levelDownsamples: number[] }
export interface InterpretationSlide extends InterpretationSlideInput, SlideGeometry {
  patchCount: number; dimensions: number; dtype: string; patchWidthLevel0: number; patchHeightLevel0: number;
  alignment: 'embedded_verified' | 'user_confirmed';
}
export interface InterpretationSelection {
  name: string; predictorId: string; evaluationId?: string | null; clinicalAnalysisId?: string | null;
  encoderId: string; slides: InterpretationSlideInput[]; resources?: ResourcePolicy;
}
export interface InterpretationSlideResult {
  slideId: string; patchCount: number; probabilities: number[]; attentionArtifact: string;
  members: { index: number; checkpointSha256: string; probabilities: number[] }[];
}
export interface InterpretationExecution extends ComputeExecution {
  result?: { slides?: InterpretationSlideResult[]; classOrder?: string[]; [key: string]: unknown } | null;
}
export interface Interpretation {
  id: string; createdAt: string; contentHash: string; lifecycleState?: LifecycleState;
  manifest: Omit<InterpretationSelection, 'slides'> & { kind: 'model-interpretation'; experimentId: string; slides: InterpretationSlide[]; method: 'ensemble' | 'refit'; memberCount: number; [key: string]: unknown };
  execution?: InterpretationExecution;
}
export interface InterpretationPreview { canSave: boolean; previewHash: string | null; findings: Finding[]; manifest: Interpretation['manifest'] | null }
export interface AttentionPatch { index: number; x: number; y: number; weight: number; percentile: number }
export interface RankedAttentionPatch extends AttentionPatch { rank: number }
export interface TopAttentionMap extends Omit<AttentionMap, 'patches'> { patches: RankedAttentionPatch[]; scope: 'whole_slide'; returned: number; coordinateBounds?: SlideRegion | null }
export interface AttentionMap {
  slideId: string; total: number; offset: number; limit: number; patchWidthLevel0: number; patchHeightLevel0: number;
  coordinateSpace: 'level0'; attentionKind: 'class_independent_pooling'; member: string | number;
  patches: AttentionPatch[]; probabilities: number[]; classOrder: string[];
}
export interface SlideRegion { x: number; y: number; width: number; height: number }
export interface InterpretationSource {
  id: string; name: string; current: boolean; findings: Finding[]; encoderId: string | null; dimensions: number | null;
  dtype: string | string[] | null; slideCount: number; featureSetId: string;
  datasetId?: string; datasetName?: string; slideFolder?: string | null; slideFolderSource?: 'dataset_import' | 'dataset_records' | null; slideFolderFinding?: Finding | null;
  packs: { id: string; name: string; outputDtype: string }[];
}
export interface GallerySource { slideFolder: string; featureBundleId: string; packArtifactId?: string | null; predictorId: string }
export interface GallerySlide { slideId: string; slidePath: string; relativePath: string; name: string; available: boolean; reason: string | null; patchCount: number | null }
export interface SlideGallery {
  items: GallerySlide[]; total: number; offset: number; limit: number; hasMore: boolean; folder: string;
  source: { featureBundleId: string; packArtifactId: string | null; encoderId: string; dimensions: number; dtype: string | string[] }; warnings: string[];
}
export interface VisualizeSelection extends GallerySource {
  slidePaths: string[]; evaluationId?: string | null; clinicalAnalysisId?: string | null; resources?: ResourcePolicy;
  patchWidthLevel0?: number; patchHeightLevel0?: number;
}
export interface VisualizeItem { slidePath: string; slideId: string; interpretationId?: string | null; status: string; reused: boolean; error?: { code: string; message: string } | null; interpretation?: Interpretation }
export interface VisualizeResult { items: VisualizeItem[]; interpretations: Interpretation[] }
const base = (project: string) => `/projects/${encodeURIComponent(project)}/interpretations`;
const record = (project: string, id: string) => `${base(project)}/${encodeURIComponent(id)}`;
const slide = (project: string, id: string, slideId: string) => `${record(project, id)}/slides/${encodeURIComponent(slideId)}`;
const post = (value: unknown) => ({ method: 'POST', body: JSON.stringify(value) });
export const interpretations = {
  sources: (project: string) => request<{ items: InterpretationSource[] }>(`${base(project)}/sources`),
  gallery: (project: string, source: GallerySource, search: string, offset: number, signal?: AbortSignal) => request<SlideGallery>(`${base(project)}/gallery`, { ...post({ ...source, search, offset, limit: 24 }), signal }),
  galleryThumbnail: (project: string, path: string, signal?: AbortSignal) => fetchArtifactBlob(`${base(project)}/gallery/thumbnail?${new URLSearchParams({ path, max_size: '320' })}`, signal),
  visualize: (project: string, selection: VisualizeSelection, operationId: string) => request<VisualizeResult>(`${base(project)}/visualize`, post({ ...selection, operationId })),
  list: (project: string) => request<{ items: Interpretation[]; executionNote?: string }>(base(project)),
  get: (project: string, id: string) => request<Interpretation>(record(project, id)),
  inspectSlide: (project: string, path: string) => request<SlideGeometry>(`${base(project)}/slide-inspection`, post({ path })),
  preview: (project: string, selection: InterpretationSelection) => request<InterpretationPreview>(`${base(project)}/preview`, post(selection)),
  save: (project: string, selection: InterpretationSelection, previewHash: string, operationId: string) => request<Interpretation>(base(project), post({ ...selection, previewHash, operationId })),
  execution: (project: string, id: string) => request<InterpretationExecution>(`${record(project, id)}/execution`),
  job: (project: string, id: string, action: 'launch' | 'resume' | 'cancel', operationId: string) => request<InterpretationExecution>(`${record(project, id)}/${action}`, post({ operationId })),
  attention: (project: string, id: string, slideId: string, member: string, region: SlideRegion, offset = 0, limit = 10000, signal?: AbortSignal) => request<AttentionMap>(`${slide(project, id, slideId)}/attention?${new URLSearchParams({ member, offset: String(offset), limit: String(limit), ...Object.fromEntries(Object.entries(region).map(([key, value]) => [key, String(value)])) })}`, { signal }),
  topAttention: (project: string, id: string, slideId: string, member: string, limit: 10 | 20, signal?: AbortSignal) => request<TopAttentionMap>(`${slide(project, id, slideId)}/attention/top?${new URLSearchParams({ member, limit: String(limit) })}`, { signal }),
  patchImage: (project: string, id: string, slideId: string, member: string, patchIndex: number, signal?: AbortSignal) => fetchArtifactBlob(`${slide(project, id, slideId)}/patches/${encodeURIComponent(patchIndex)}/image?${new URLSearchParams({ member, max_size: '512' })}`, signal),
  thumbnail: (project: string, id: string, slideId: string, signal?: AbortSignal) => fetchArtifactBlob(`${slide(project, id, slideId)}/thumbnail?max_size=1536`, signal),
  region: (project: string, id: string, slideId: string, region: SlideRegion, signal?: AbortSignal) => fetchArtifactBlob(`${slide(project, id, slideId)}/region?${new URLSearchParams({ ...Object.fromEntries(Object.entries(region).map(([key, value]) => [key, String(value)])), max_size: '1536' })}`, signal),
  download: (project: string, id: string, filename: string) => downloadArtifact(`${record(project, id)}/artifacts/${encodeURIComponent(filename)}`, filename),
};

File: histopilot/api/interpretation.py
"""Attention study review, durable execution, authenticated slide views and exports."""

from fastapi import APIRouter, Path, Query, Response

from histopilot.application.interpretation import InterpretationService
from histopilot.schemas.interpretation import (
    InterpretationGalleryQuery,
    InterpretationSelection,
    SaveInterpretation,
    SlideInspection,
    VisualizeInterpretation,
)
from histopilot.schemas.predictors import PredictorAction
from histopilot.storage.project_lock import StorageError


def interpretation_router(projects, filesystem):
    router = APIRouter(prefix="/api/v1/projects/{identity}/interpretations")

    def service(identity):
        return InterpretationService(projects.scientific_store(identity), filesystem)

    @router.get("")
    def list_interpretations(identity: str, include_inactive: bool = False):
        return service(identity).list(include_inactive=include_inactive)

    @router.get("/sources")
    def sources(identity: str):
        return service(identity).gallery.sources()

    @router.post("/gallery")
    def gallery(identity: str, payload: InterpretationGalleryQuery):
        return service(identity).gallery.query(payload)

    @router.get("/gallery/thumbnail")
    def gallery_thumbnail(
        identity: str,
        path: str = Query(min_length=1, max_length=4096),
        max_size: int = Query(default=256, ge=64, le=512),
    ):
        return Response(
            content=service(identity).gallery_thumbnail(path, max_size=max_size),
            media_type="image/png",
            headers={"Cache-Control": "private, no-store"},
        )

    @router.post("/visualize", status_code=202)
    def visualize(identity: str, payload: VisualizeInterpretation):
        return service(identity).visualize(payload)

    @router.post("/slide-inspection")
    def inspect_slide(identity: str, payload: SlideInspection):
        return service(identity).inspect_slide(payload.path)

    @router.post("/preview")
    def preview(identity: str, payload: InterpretationSelection):
        return service(identity).preview(payload)

    @router.post("", status_code=201)
    def save(identity: str, payload: SaveInterpretation):
        return service(identity).save(payload)

    @router.get("/{interpretation_id}")
    def get(identity: str, interpretation_id: str):
        return service(identity).get(interpretation_id)

    @router.get("/{interpretation_id}/execution")
    def execution(identity: str, interpretation_id: str):
        return service(identity).execution(interpretation_id)

    @router.post("/{interpretation_id}/launch", status_code=202)
    def launch(identity: str, interpretation_id: str, payload: PredictorAction):
        return service(identity).launch(interpretation_id, payload.operationId)

    @router.post("/{interpretation_id}/resume", status_code=202)
    def resume(identity: str, interpretation_id: str, payload: PredictorAction):
        return service(identity).launch(interpretation_id, payload.operationId, resume=True)

    @router.post("/{interpretation_id}/cancel", status_code=202)
    def cancel(identity: str, interpretation_id: str, payload: PredictorAction):
        return service(identity).cancel(interpretation_id, payload.operationId)

    @router.get("/{interpretation_id}/artifacts/{filename}")
    def artifact(identity: str, interpretation_id: str, filename: str):
        return Response(
            content=service(identity).artifact(interpretation_id, filename),
            media_type="application/octet-stream"
            if filename.endswith(".npy")
            else "application/json",
            headers={"Content-Disposition": f'attachment; filename="{filename}"'},
        )

    @router.get("/{interpretation_id}/slides/{slide_id}/thumbnail")
    def thumbnail(
        identity: str,
        interpretation_id: str,
        slide_id: str,
        max_size: int = Query(default=1024, ge=64, le=2048),
    ):
        return Response(
            content=service(identity).image(interpretation_id, slide_id, max_size=max_size),
            media_type="image/png",
            headers={"Cache-Control": "private, no-store"},
        )

    @router.get("/{interpretation_id}/slides/{slide_id}/region")
    def region(
        identity: str,
        interpretation_id: str,
        slide_id: str,
        x: int = Query(ge=0),
        y: int = Query(ge=0),
        width: int = Query(gt=0),
        height: int = Query(gt=0),
        max_size: int = Query(default=1024, ge=64, le=2048),
    ):
        return Response(
            content=service(identity).image(
                interpretation_id, slide_id, max_size=max_size, region=(x, y, width, height)
            ),
            media_type="image/png",
            headers={"Cache-Control": "private, no-store"},
        )

    @router.get("/{interpretation_id}/slides/{slide_id}/attention")
    def attention(
        identity: str,
        interpretation_id: str,
        slide_id: str,
        member: str = "mean",
        offset: int = Query(default=0, ge=0),
        limit: int = Query(default=10000, ge=1, le=50000),
        x: int | None = Query(default=None, ge=0),
        y: int | None = Query(default=None, ge=0),
        width: int | None = Query(default=None, gt=0),
        height: int | None = Query(default=None, gt=0),
    ):
        bounds = (x, y, width, height)
        if any(value is not None for value in bounds) and not all(
            value is not None for value in bounds
        ):
            raise StorageError(
                "Provide all four viewport bounds.", "INTERPRETATION_VIEW_INVALID", 422
            )
        return service(identity).attention(
            interpretation_id,
            slide_id,
            member=member,
            offset=offset,
            limit=limit,
            region=bounds if x is not None else None,
        )

    @router.get("/{interpretation_id}/slides/{slide_id}/attention/top")
    def top_attention(
        identity: str,
        interpretation_id: str,
        slide_id: str,
        member: str = "mean",
        limit: int = Query(default=10, ge=1, le=20),
    ):
        return service(identity).top_attention(
            interpretation_id, slide_id, member=member, limit=limit
        )

    @router.get("/{interpretation_id}/slides/{slide_id}/patches/{patch_index}/image")
    def patch_image(
        identity: str,
        interpretation_id: str,
        slide_id: str,
        patch_index: int = Path(ge=0),
        member: str = "mean",
        max_size: int = Query(default=512, ge=64, le=1024),
    ):
        return Response(
            content=service(identity).patch_image(
                interpretation_id, slide_id, patch_index, member=member, max_size=max_size
            ),
            media_type="image/png",
            headers={"Cache-Control": "private, no-store"},
        )

    return router
